File: 2 - Work/Phase 3 Economic Merge and MICE Imputation/Bulk Tests/python/rubins_rules_pooling.py
# Purpose: Pool the 5 MICE-imputed aggregate estimates using Rubin's Rules
#          to produce a single national land-value point estimate with 95% CI.
# Inputs:  Bulk Tests/python/Py_Imputed_Dataset_{1..5}.csv
# Outputs: Bulk Tests/python/Py_Rubins_Rules_Summary.csv


# === 1. LIBRARIES ===

import pathlib

import numpy as np
import pandas as pd


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR = pathlib.Path(__file__).parent
IN_DIR     = SCRIPT_DIR
OUT_CSV    = SCRIPT_DIR / "Py_Rubins_Rules_Summary.csv"

M = 5


# === 3. FUNCTIONS ===

def run_pooling(in_dir, out_csv, m_datasets=5):
    aggregates  = []
    within_vars = []

    print("--- 1  Loading imputed datasets and computing aggregates ---\n")
    for i in range(1, m_datasets + 1):
        fpath = in_dir / f"Py_Imputed_Dataset_{i}.csv"

        if not fpath.exists():
            raise FileNotFoundError(f"Imputed dataset not found: {fpath}")

        df = pd.read_csv(fpath)
        df["Total_Opportunity_Cost"] = (
            df["osm_acreage"] * df["Baseline_Value_Per_Acre"]
        )

        q_i   = df["Total_Opportunity_Cost"].sum()
        var_i = df["Total_Opportunity_Cost"].var()

        aggregates.append(q_i)
        within_vars.append(var_i)

        print(f"  Dataset {i}:  ${q_i / 1e9:>10,.3f} B")

    aggregates  = np.array(aggregates)
    within_vars = np.array(within_vars)

    # [METHODOLOGY] Rubin's Rules pooling — q_bar is the pooled national estimate;
    #               v_t combines within- and between-imputation variance (Rubin 1987)
    print("\n--- 2  Applying Rubin's Rules ---")

    q_bar = aggregates.mean()
    v_w   = within_vars.mean()
    v_b   = aggregates.var(ddof=1)
    v_t   = v_w + v_b + v_b / m_datasets
    se    = np.sqrt(v_t)
    ci_lo = q_bar - 2.576 * se
    ci_hi = q_bar + 2.576 * se

    print(f"\n=== RUBIN'S RULES RESULTS ===")
    print(f"  Pooled Aggregate National Value:  ${q_bar / 1e9:,.3f} Billion")
    print(f"  Within-Imputation Variance (v_w): {v_w:.4e}")
    print(f"  Between-Imputation Variance (v_b):{v_b:.4e}")
    print(f"  Total Variance (v_t):             {v_t:.4e}")
    print(f"  Standard Error:                   ${se / 1e9:,.3f} Billion")
    print(
        f"  99% Confidence Interval:          "
        f"${ci_lo / 1e9:,.3f} B  -  ${ci_hi / 1e9:,.3f} B"
    )

    pooled_df = pd.DataFrame({
        "Metric": [
            "Pooled Aggregate National Value ($)",
            "Pooled Aggregate National Value ($B)",
            "Within-Imputation Variance (v_w)",
            "Between-Imputation Variance (v_b)",
            "Total Variance (v_t)",
            "Standard Error ($)",
            "95% CI Lower ($B)",
            "95% CI Upper ($B)",
        ] + [f"Dataset {i} Aggregate ($B)" for i in range(1, m_datasets + 1)],
        "Value": [
            f"{q_bar:,.2f}",
            f"{q_bar / 1e9:,.3f}",
            f"{v_w:.4e}",
            f"{v_b:.4e}",
            f"{v_t:.4e}",
            f"{se:,.2f}",
            f"{ci_lo / 1e9:,.3f}",
            f"{ci_hi / 1e9:,.3f}",
        ] + [f"{q / 1e9:,.3f}" for q in aggregates],
    })

    pooled_df.to_csv(out_csv, index=False)
    print(f"\n  [OK] Saved -> {out_csv}")


# === 4. EXECUTION ===

if __name__ == "__main__":
    run_pooling(IN_DIR, OUT_CSV, M)

File: 2 - Work/Phase 3 Economic Merge and MICE Imputation/Bulk Tests/R/rubins_rules_pooling.R
# Purpose: Pool the 5 MICE-imputed aggregate estimates using Rubin's Rules
#          to produce a single national land-value point estimate with 95% CI.
# Inputs:  Bulk Tests/R/R_Imputed_Dataset_{1..5}.csv
# Outputs: Bulk Tests/R/R_Rubins_Rules_Summary.csv


# === 1. LIBRARIES ===

suppressPackageStartupMessages({
  library(tidyverse)
  library(future)
  library(furrr)
  library(parallelly)
  library(this.path)
})


# === 2. GLOBALS & PATHS ===

SCRIPT_DIR  <- this.path::this.dir()
OUT_CSV     <- file.path(SCRIPT_DIR, "R_Rubins_Rules_Summary.csv")

M            <- 5
IMPUTED_DIR  <- SCRIPT_DIR

# Reserve fewer cores than MICE.R — pooling is sequential, workers only kept
# for consistency with the parallel environment setup
SAFE_WORKERS <- min(availableCores() - 12, 20)
plan(multisession, workers = SAFE_WORKERS)


# === 3. FUNCTIONS ===

#' Apply Rubin's Rules to m point estimates and within-dataset variances.
#' @param Q_vec Numeric vector of length m (aggregate Q_i per dataset).
#' @param U_vec Numeric vector of length m (within-dataset variance U_i).
#' @return Named list with Q_bar, V_W, V_B, V_T, SE, CI_lo, CI_hi.
rubins_rules <- function(Q_vec, U_vec = NULL) {
  m     <- length(Q_vec)
  Q_bar <- mean(Q_vec)
  V_B   <- var(Q_vec)
  V_W   <- if (!is.null(U_vec) && length(U_vec) == m) mean(U_vec) else 0
  V_T   <- V_W + V_B + V_B / m
  SE    <- sqrt(V_T)
  list(
    Q_bar = Q_bar, V_W = V_W, V_B = V_B, V_T = V_T,
    SE    = SE,
    CI_lo = Q_bar - 2.576 * SE,
    CI_hi = Q_bar + 2.576 * SE,
    m     = m
  )
}


# === 4. EXECUTION ===

main <- function() {
  aggregates  <- numeric(M)
  within_vars <- numeric(M)

  cat("--- 1  Loading imputed datasets and computing aggregates ---\n\n")

  for (i in 1:M) {
    filepath <- file.path(IMPUTED_DIR, sprintf("R_Imputed_Dataset_%d.csv", i))

    if (!file.exists(filepath)) {
      stop(sprintf("File not found: %s", filepath))
    }

    df <- read_csv(filepath, show_col_types = FALSE)

    df$Total_Opportunity_Cost <- df[["osm_acreage"]] * df[["Baseline_Value_Per_Acre"]]

    Q_i   <- sum(df$Total_Opportunity_Cost, na.rm = TRUE)
    Var_i <- var(df$Total_Opportunity_Cost, na.rm = TRUE)

    aggregates[i]  <- Q_i
    within_vars[i] <- Var_i

    cat(sprintf("  Dataset %d:  $%10.3f B\n", i, Q_i / 1e9))
  }

  cat("\n--- 2  Applying Rubin's Rules ---\n")

  # [METHODOLOGY] Rubin's Rules pooling — Q_bar is the pooled national estimate;
  #               V_T combines within- and between-imputation variance (Rubin 1987)
  pool   <- rubins_rules(aggregates, within_vars)
  Q_bar  <- pool$Q_bar
  V_W    <- pool$V_W
  V_B    <- pool$V_B
  V_T    <- pool$V_T
  SE     <- pool$SE
  CI_lo  <- pool$CI_lo
  CI_hi  <- pool$CI_hi

  cat("\n=== RUBIN'S RULES RESULTS ===\n")
  cat(sprintf("  Pooled Aggregate National Value:  $%10.3f B\n", Q_bar / 1e9))
  cat(sprintf("  Within-Imputation Variance (V_W): %.4e\n", V_W))
  cat(sprintf("  Between-Imputation Variance (V_B):%.4e\n", V_B))
  cat(sprintf("  Total Variance (V_T):             %.4e\n", V_T))
  cat(sprintf("  Standard Error:                   $%10.3f B\n", SE / 1e9))
  cat(sprintf(
    "  95%% Confidence Interval:          $%10.3f B - $%10.3f B\n",
    CI_lo / 1e9, CI_hi / 1e9
  ))

  pooled_df <- data.frame(
    Metric = c(
      "Pooled Aggregate National Value ($)",
      "Pooled Aggregate National Value ($B)",
      "Within-Imputation Variance (V_W)",
      "Between-Imputation Variance (V_B)",
      "Total Variance (V_T)",
      "Standard Error ($)",
      "95% CI Lower ($B)",
      "95% CI Upper ($B)",
      paste0("Dataset ", 1:M, " Aggregate ($B)")
    ),
    Value = c(
      format(Q_bar, big.mark = ",", scientific = FALSE),
      sprintf("%.3f", Q_bar / 1e9),
      sprintf("%.4e", V_W),
      sprintf("%.4e", V_B),
      sprintf("%.4e", V_T),
      format(SE, big.mark = ",", scientific = FALSE),
      sprintf("%.3f", CI_lo / 1e9),
      sprintf("%.3f", CI_hi / 1e9),
      sprintf("%.3f", aggregates / 1e9)
    )
  )

  write_csv(pooled_df, OUT_CSV)
  cat(sprintf("\n  [OK] Saved -> %s\n", OUT_CSV))
}

main()
